File: PRACTICA 5/eventos.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include<math.h>
#include <time.h>

struct Evento {
    int inicio;
    int fin;
    int duracion;
};

void random_evento(struct Evento* evento, int max_inicio, int max_duracion) {
    srand(time(NULL) + rand());
    evento->inicio = rand() % max_inicio;
    evento->duracion = (rand() % max_duracion) + 1; 
    evento->fin = evento->inicio + evento->duracion;
}

int maximo_eventos(struct Evento eventos[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (eventos[j].fin > eventos[j + 1].fin) {
                struct Evento temp = eventos[j];
                eventos[j] = eventos[j + 1];
                eventos[j + 1] = temp;
            }
        }
    }
    int contador = 0;
    int ultimo_fin = -1;

    for (int i = 0; i < n; i++) {
        if (eventos[i].inicio >= ultimo_fin) {
            contador++;
            ultimo_fin = eventos[i].fin;
        }
    }
    return contador;
}

int main(int argc, char *argv[]) {
    int n_eventos = 10;
    struct Evento eventos[n_eventos];

    for (int i = 0; i < n_eventos; i++) {
        random_evento(&eventos[i], 12, 6);
        printf("Evento %d: Inicio: %d, Fin: %d, Duracion: %d\n", i + 1, eventos[i].inicio, eventos[i].fin, eventos[i].duracion);
    }

    int max_eventos = maximo_eventos(eventos, n_eventos);
    printf("Numero maximo de eventos: %d\n", max_eventos);

    system("pause");
    return 0;
    
}

File: PRACTICA 5/eventos.py
from datetime import datetime, timedelta

class Evento:
    def __init__(self, fecha_inicio, fecha_fin):
        self.fecha_inicio = fecha_inicio
        self.fecha_fin = fecha_fin
    def duracion(self):
        return self.fecha_fin - self.fecha_inicio
    

def random_eventos(n, inicio, fin, duracion_min, duracion_max):
    import random
    eventos = []
    delta = fin - inicio
    for _ in range(n):
        inicio_evento = inicio + timedelta(hours=random.randint(0, int(delta.total_seconds() // 3600)))
        duracion_evento = timedelta(hours=random.randint(duracion_min, duracion_max))
        fin_evento = inicio_evento + duracion_evento
        eventos.append(Evento(inicio_evento, fin_evento))
    return eventos

def maximo_eventos(eventos):
    eventos_ordenados = sorted(eventos, key=lambda evento: evento.fecha_fin) 
    contador = 0
    ultimo_fin = datetime.min


    for evento in eventos_ordenados:
        if evento.fecha_inicio >= ultimo_fin:
            contador += 1
            ultimo_fin = evento.fecha_fin

    return contador


if __name__ == "__main__":
    inicio = datetime(2025, 11, 1)
    fin = datetime(2025, 11, 3)
    eventos = random_eventos(5, inicio, fin, 1, 6) 
    for i, evento in enumerate(eventos):
        print(f"Evento {i+1}: Inicio: {evento.fecha_inicio}, Fin: {evento.fecha_fin}, Duración: {evento.duracion()}")

    max_eventos = maximo_eventos(eventos)
    print(f"Número máximo de eventos que se pueden asistir: {max_eventos}")
